### PR TITLE
Invoke ansi formatting when a IPython shell exists

### LINKED ISSUE #115: IPython extension causes confusing help text?
If I'm using ipython and I run ``%load_ext param.ipython`` everything seems good. But if I'm not using ipython, or if I use ipython without loading the ipython extension, I think the help text for a parameterized instance is confusing.

Normal python:

```python
>>> import param
>>> class X(param.Parameterized):
...     y = param.Parameter(default=1)
... 
>>> x = X(y=10)
>>> help(x)
```

Gives me:
```
Help on X in module __main__ object:

class X(param.parameterized.Parameterized)
 |  params(y=Parameter, name=String)
 |  ESC[1;32mParameters of 'X'
 |  =================
 |  ESC[0m
 |  ESC[1;31mParameters changed from their default values are marked in red.ESC[0m
 |  ESC[1;36mSoft bound values are marked in cyan.ESC[0m
 |  C/V= Constant/Variable, RO/RW = ReadOnly/ReadWrite, AN=Allow None
 |  
 |  ESC[1;34mNameValue    Type    Mode ESC[0m
 |  
 |  y   1  Parameter  V RW 
 |  
 |  ESC[1;32mParameter docstrings:
 |  =====================ESC[0m
 |  
 |  ESC[1;34my: < No docstring available >ESC[0m
 |  
...
```

Maybe my terminal's not very clever, so it can't display colors, but you can see that there was no attempt to color the ``y`` line, and the value shown is the default rather than the actual value
ipython
.

Additionally, the attached screenshot shows the result from within ipython (without having run ``%load_ext param.ipython``).

### PR DESCRIPTION
**Pull request by tonyfast**
fixes #115

### CODE REVIEW

**Comment by jlstevens** [comment]
Looks great though something weird happened in the docstring.

File: param/ipython.py [hunk 1]
```diff
@@ -55,7 +58,7 @@ def __init__(self, metaclass=False):
     def get_param_info(self, obj, include_super=True):
         """
         Get the parameter dictionary, the list of modifed parameters
-        and the dictionary or parameter values. If include_super is
+        and the dictionary or parameter values. If include_s'https://tools.ietf.org/html/rfc7946'uper is
```
Comment (jbednar) [inline param/ipython.py:61]:
```suggestion
        and the dictionary or parameter values. If include_super is
```

**Comment by tonyfast** [comment]
thanks for cleaning that up @jbednar bednar. sorry i missed it.

**Comment by jbednar** [comment]
This looks good, but I can't tell whether it's the right thing to do. This file is called `ipython.py`, indicating to me that the entire file is about customizations specifically for ipython, but it sounds like it's used even when IPython is not in use (or else such changes wouldn't be needed?). I find that very confusing. `parameterized.py` loads it using a try/except, and seems to have code already for working around the absence of that file, but I think that's meant for when Param is vendored into another codebase.  If there is code in ipython.py that's meant to be usable with or without IPython, it seems to me it should move out of this file, rather than have this file have workarounds for when IPython is not in use. @jlstevens ?

**Comment by jlstevens** [comment]
> If there is code in ipython.py that's meant to be usable with or without IPython, it seems to me it should move out of this file, rather than have this file have workarounds for when IPython is not in use.

A good bit of the code here is about generating param docs with ANSI color codes. In that sense, it is not ipython specific but ipython will be the most common environment where such colors can be nicely displayed. HoloViews uses this mechanism in `hv.help` to display information about plot options but I *believe* that will fall back to the normal `help` if it needs to (i.e if this file is unavailable). The issue is that you want to use the colors but only when they are supported, or you see the ANSI codes.

I would rather not add more files to param and I don't think it is a huge deal to have this functionality in a file called `ipython.py`. If this file was purely ipython specific, it would be tiny and there would be a chunk of code that would have to live somewhere else.

**Comment by jbednar** [comment]
I'm just not seeing what this code has to do with IPython.  Is this using IPython somehow, even though I'm invoking it as just `python`?

image

It doesn't seem like checking for IPython is the right way to control the ANSI codes.

**Comment by jbednar** [comment]
https://stackoverflow.com/questions/7445658/how-to-detect-if-the-console-does-support-ansi-escape-codes-in-python ?

https://bixense.com/clicolors/?

**Comment by jlstevens** [comment]
I would vote for option 2 and docs to explain the presence of the flag when ANSI is not supported.

**Comment by tonyfast** [comment]
closing because this not the solution.

**Pull request closed without merging**